// @microsoft/teamsfx-core/build/component/debugHandler/error.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidTabDebugArgsError = exports.BotMessagingEndpointMissingError = exports.InvalidExistingBotArgsError = exports.InvalidSSODebugArgsError = exports.InvalidAppManifestPackageFileFormatError = exports.AppManifestPackageNotExistError = exports.errorSource = void 0;
const tslib_1 = require("tslib");
const util = tslib_1.__importStar(require("util"));
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const localizeUtils_1 = require("../../common/localizeUtils");
exports.errorSource = "debugHandler";
function AppManifestPackageNotExistError(appManifestPackagePath) {
    return new teamsfx_api_1.UserError(exports.errorSource, "InvalidAppManifestDebugArgsError", util.format(localizeUtils_1.getDefaultString("error.debugHandler.AppManifestPackageNotExistError"), appManifestPackagePath), util.format(localizeUtils_1.getLocalizedString("error.debugHandler.AppManifestPackageNotExistError"), appManifestPackagePath));
}
exports.AppManifestPackageNotExistError = AppManifestPackageNotExistError;
function InvalidAppManifestPackageFileFormatError() {
    return new teamsfx_api_1.UserError(exports.errorSource, "InvalidAppManifestDebugArgsError", localizeUtils_1.getDefaultString("error.debugHandler.InvalidAppManifestPackageFileFormatError"), localizeUtils_1.getLocalizedString("error.debugHandler.InvalidAppManifestPackageFileFormatError"));
}
exports.InvalidAppManifestPackageFileFormatError = InvalidAppManifestPackageFileFormatError;
function InvalidSSODebugArgsError() {
    return new teamsfx_api_1.UserError(exports.errorSource, "InvalidSSODebugArgsError", localizeUtils_1.getDefaultString("error.debugHandler.InvalidSSODebugArgsError"), localizeUtils_1.getLocalizedString("error.debugHandler.InvalidSSODebugArgsError"));
}
exports.InvalidSSODebugArgsError = InvalidSSODebugArgsError;
function InvalidExistingBotArgsError() {
    return new teamsfx_api_1.UserError(exports.errorSource, "InvalidBotDebugArgsError", localizeUtils_1.getDefaultString("error.debugHandler.InvalidExistingBotArgsError"), localizeUtils_1.getLocalizedString("error.debugHandler.InvalidExistingBotArgsError"));
}
exports.InvalidExistingBotArgsError = InvalidExistingBotArgsError;
function BotMessagingEndpointMissingError() {
    return new teamsfx_api_1.UserError(exports.errorSource, "InvalidBotDebugArgsError", localizeUtils_1.getDefaultString("error.debugHandler.BotMessagingEndpointMissingError"), localizeUtils_1.getLocalizedString("error.debugHandler.BotMessagingEndpointMissingError"));
}
exports.BotMessagingEndpointMissingError = BotMessagingEndpointMissingError;
function InvalidTabDebugArgsError() {
    return new teamsfx_api_1.UserError(exports.errorSource, "InvalidTabDebugArgsError", localizeUtils_1.getDefaultString("error.debugHandler.InvalidTabDebugArgsError"), localizeUtils_1.getLocalizedString("error.debugHandler.InvalidTabDebugArgsError"));
}
exports.InvalidTabDebugArgsError = InvalidTabDebugArgsError;
